Add 2 metric tags for metric total.build.time

zhaoamong opened the pull request:
Hi Vijay,
Thanks for your time reviewing this code change. This is a small change: Justin agreed to let me add 2 tags for the metric named "total.build.time":
git_commit
build_number
And this is the code change and unit test for it.

With these 2 additional tags I can get all the necessary info to develop our CI Service visibility report for CRM Infra.

Thanks a lot!

VijayKumarMidde commented on src/main/java/org/jenkinsci/plugins/argusnotifier/TagFactory.java:
```diff
@@ -39,6 +40,20 @@ private TagFactory() {
+                .putAll(hostTag(jenkins))
+                .put(PROJECT.lower(), InvalidCharSwap.swapWithDash(projectName));
+                
+        if (!commitId.isEmpty()) {
```
Is this check needed?  `buildExtendedStatusTags` never gets called if `commitId` is empty.
https://github.com/jenkinsci/argus-notifier-plugin/blob/574c866442ffb6075056bd6968e18d50f344c447/src/main/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactory.java#L99-L101

zhaoamong replied:
This is defensive check in TagFactory. But I agree it is duplicate check. I removed this check.

VijayKumarMidde commented on src/main/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatter.java:
```diff
@@ -6,11 +6,13 @@
  */
 class JenkinsRunFormatter {
     
+	public static final String GIT_COMMIT = "GIT_COMMIT";
```
Can you please format all of the source files that you have modified? Looks like there is mismatch with indentation settings.

zhaoamong replied:
fixed

zhaoamong replied:
Use the IDE formatter to replaced TAB with spaces

VijayKumarMidde commented on src/main/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactory.java:
```diff
@@ -95,8 +95,15 @@ private Metric getBuildTimeMetric(String labelForDisplayName, String metricStrin
         double timeInSeconds = (double) timeInMillis / 1000.0;
 
-        metric.setTags(TagFactory.buildStatusTags(jenkins,
+        String gitCommitId = jenkinsRunFormatter.getGitCommit();
```
Can you please add a test case for this change?

zhaoamong replied:
Yes, added one test case to cover gitCommit is a valid string and empty.

VijayKumarMidde replied:
Thank you!

VijayKumarMidde commented on pom.xml:
```diff
@@ -1,27 +1,25 @@
     <properties>
         <!-- Baseline Jenkins version you use to build the plugin. Users must have this version or newer to run. -->
-        <jenkins.version>2.60</jenkins.version>
+        <jenkins.version>2.164.3</jenkins.version>
```
Please change this to out current supported version `2.164.2.1`

zhaoamong replied:
done

justinharringa replied:
Just an FYI, you want to be careful about how far you bump these versions in case folks in the open-source community haven't updated. In this case, this should be fine since this version has been around since May.

justinharringa commented on pom.xml:
```diff
@@ -1,27 +1,25 @@
-        <relativePath />
+      <groupId>org.jenkins-ci.plugins</groupId>
+      <artifactId>plugin</artifactId>
+      <version>3.43</version> <!-- or whatever the newest version available is -->
```
Why the comment?

zhaoamong replied:
removed comment. Have to use a newer version 3.43 to build the project.

zhaoamong commented on pom.xml:
```diff
@@ -1,27 +1,25 @@
         <java.version>1.8</java.version>
         <!-- Jenkins Test Harness version you use to test the plugin. -->
         <!-- For Jenkins version >= 1.580.1 use JTH 2.x or higher. -->
-        <jenkins-test-harness.version>2.13</jenkins-test-harness.version>
```
added back, updated to 2.49

zhaoamong commented on pom.xml:
```diff
@@ -1,27 +1,25 @@
+    <version>1.0.7-SNAPSHOT</version>
     <packaging>hpi</packaging>
 
     <properties>
```
done

zhaoamong commented on src/test/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatterTest.groovy:
```diff
@@ -70,6 +71,43 @@ class JenkinsRunFormatterTest extends Specification {
+        actualGitCommit == expectedGitCommit
+
+        where:
+        commitId               | expectedGitCommit
```
done

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -43,4 +48,36 @@ class BuildMetricFactoryTest extends Specification {
+        String jobFolderName = "a_jobFolderName"
+        folder.fullName >> jobFolderName
+        String commitTag = "GIT_COMMIT"
+        ImmutableMap.Builder<String, String> mapBuilder = ImmutableMap.<String, String>builder().put(commitTag,commitId);
```
Semicolons are unnecessary and inconsistent in several spots.

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -43,4 +48,36 @@ class BuildMetricFactoryTest extends Specification {
+        folder.fullName >> jobFolderName
+        String commitTag = "GIT_COMMIT"
+        ImmutableMap.Builder<String, String> mapBuilder = ImmutableMap.<String, String>builder().put(commitTag,commitId);
+        Map<String, String> envVars = mapBuilder.build();
```
No need for the builder here and you can use constants instead of needing `commitTag` on line 60 (this change removes the need for the other 2 lines above).
```suggestion
        Map<String, String> envVars = [(TagFactory.Tag.GIT_COMMIT.toString()): commitId]
```

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -43,4 +48,36 @@ class BuildMetricFactoryTest extends Specification {
+        BuildMetricFactory metricFactory = new BuildMetricFactory(jenkins, run, metricTimestamp, scope)
+
+        when:
+        Metric metricTotalTime = metricFactory.getBuildTimeMetric(BuildMetricFactory.TOTAL_BUILD_TIME_LABEL,BuildMetricFactory.TOTAL_BUILD_TIME_METRIC,totalDuration)
```
This line is too long

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -43,4 +48,36 @@ class BuildMetricFactoryTest extends Specification {
+        then:
+
+        metricTotalTime.tags[commitTag.toLowerCase()] == expectedCommit
+        metricTotalTime.tags['build_number'] == commitId.isEmpty() ? null : String.valueOf(buildNumber)
```
Is this really what we'd want? It seems like the absence of `GIT_COMMIT` shouldn't influence `BUILD_NUMBER`

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -43,4 +48,36 @@ class BuildMetricFactoryTest extends Specification {
+
+        metricTotalTime.tags[commitTag.toLowerCase()] == expectedCommit
+        metricTotalTime.tags['build_number'] == commitId.isEmpty() ? null : String.valueOf(buildNumber)
+        metricTotalTime.tags['project'].startsWith(jobFolderName)
```
Is this related to what we want to test here? Seems unnecessary.

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -43,4 +48,36 @@ class BuildMetricFactoryTest extends Specification {
+        where:
+        commitId                  | expectedCommit
+        "32ds34frg4534ff34fdcdds" | "32ds34frg4534ff34fdcdds"
+        ""                        | null
```
Are there more tests here? Such as a `commitId` of `null`?

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -43,4 +48,36 @@ class BuildMetricFactoryTest extends Specification {
         ""            | FULL_JOB_NAME + ": " + BuildMetricFactory.NUMERIC_BUILD_STATUS_LABEL
     }
+
+    def 'Total build time metric has git commit tag'() {
```
```suggestion
    def 'Total build time metric with #commitId has git commit tag #expectedCommit'() {
```

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -1,6 +1,11 @@
 package org.jenkinsci.plugins.argusnotifier
 
+import org.apache.commons.jelly.impl.TagFactory
```
Incorrect import
```suggestion
```

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -1,6 +1,11 @@
 
+import org.apache.commons.jelly.impl.TagFactory
+
+import com.google.common.collect.ImmutableMap
```
Unnecessary import
```suggestion
```

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/BuildMetricFactoryTest.groovy:
```diff
@@ -1,6 +1,11 @@
+import com.google.common.collect.ImmutableMap
 import com.salesforce.dva.argus.sdk.entity.Metric
+
+import groovy.json.StringEscapeUtils
```
Unused import
```suggestion
```

justinharringa commented on src/main/java/org/jenkinsci/plugins/argusnotifier/TagFactory.java:
```diff
@@ -34,15 +32,21 @@ private TagFactory() {
+                .put(PROJECT.lower(), InvalidCharSwap.swapWithDash(projectName)).build();
+    }
+
+    static Map<String, String> buildExtendedStatusTags(Jenkins jenkins, String projectName, String buildNumber,
```
As mentioned previously, perhaps this is just one method with an optional `commitId`.

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatterTest.groovy:
```diff
@@ -7,6 +7,7 @@ import org.jvnet.hudson.test.JenkinsRule
 import org.jvnet.hudson.test.MockFolder
 import spock.lang.Specification
 import spock.lang.Unroll
+import com.google.common.collect.ImmutableMap
```
Unnecessary import
```suggestion
```

justinharringa commented on src/main/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatter.java:
```diff
@@ -10,7 +10,8 @@
 class JenkinsRunFormatter {
-    
+
+    public static final String GIT_COMMIT = "GIT_COMMIT";
```
This is a duplication of `TagFactory.Tag.GIT_COMMIT.toString()`

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatterTest.groovy:
```diff
@@ -70,6 +71,43 @@ class JenkinsRunFormatterTest extends Specification {
+
+        ImmutableMap.Builder<String,String> mapBuilder = ImmutableMap.<String, String>builder()
+        mapBuilder.put(JenkinsRunFormatter.GIT_COMMIT, gitCommit)
+        build.getEnvVars() >> mapBuilder.build()
```
You can squash the 2 lines above and the unnecessary import with this.
```suggestion
        build.getEnvVars() >> [(TagFactory.Tag.GIT_COMMIT.toString()): commitId]
```

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatterTest.groovy:
```diff
@@ -70,6 +71,43 @@ class JenkinsRunFormatterTest extends Specification {
 
+    def "test getGitCommit returns commit sha"() {
+        given:
+        def gitCommit = commitId
```
What is the reason for this?

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatterTest.groovy:
```diff
@@ -70,6 +71,43 @@ class JenkinsRunFormatterTest extends Specification {
+
+        ImmutableMap.Builder<String,String> mapBuilder = ImmutableMap.<String, String>builder()
+        mapBuilder.put("ENV_VAR", gitCommit)
+        build.getEnvVars() >> mapBuilder.build()
```
You can squash the above builder lines as well as `gitCommit` (not sure that it adds any value).
```suggestion
        build.getEnvVars() >> [:]
```

justinharringa commented on src/test/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatterTest.groovy:
```diff
@@ -70,6 +71,43 @@ class JenkinsRunFormatterTest extends Specification {
         actualBuildNumberString == String.valueOf(buildNumber)
     }
 
+    def "test getGitCommit returns commit sha"() {
```
```suggestion
    def "test getGitCommit for #commitId returns commit sha #expectedGitCommit"() {
```

justinharringa commented:
@zhaoamong BTW thanks for the contribution! Most of the review comments should be easily resolvable if I haven't provided feedback. To address the deprecated `getEnvVars()` method, you can use [Run.getEnvironment(TaskListener)](https://javadoc.jenkins-ci.org/hudson/model/Run.html#getEnvironment-hudson.model.TaskListener-)

Thanks again!

zhaoamong commented on src/main/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatter.java:
```diff
@@ -56,6 +57,16 @@ public String getBuildNumberString() {
 
+    public String getGitCommit() {
+
+        Object commitId = run.getEnvVars().get(GIT_COMMIT);
```
Hi @justinharringa 
For this one,
1. In existing code, "Jenkins instance = Jenkins.**getInstance**()" in ArgusRunListener.java is also deprecated. Can I follow the precedence to use getEnvVars()? 
2. I'd like to keep the code change minimum, but the new method getEnvironment(TaskListener) asks for passing in "TaskListener", then I have to touch the constructor of several classes(BuildMetricFactory,JenkinsRunFormatter etc) to pass in a "TaskListener", therefore touch more test code as well. Would this be ok with you?

justinharringa replied:
1. I really would prefer not. In the other case, it was deprecated after the code had been written. Indeed, we should squash that as well as a separate PR.

2. I don’t think that you should need to pass in a TaskListener. I believe you can use a new LogTaskListener without adverse effects. If that's not the case, we can discuss further.

zhaoamong commented on src/main/java/org/jenkinsci/plugins/argusnotifier/TagFactory.java:
```diff
@@ -38,11 +36,34 @@ private TagFactory() {
+    static Map<String, String> buildExtendedStatusTags(Jenkins jenkins, String projectName, String buildNumber,
+            String commitId, String status) {
+
+        ImmutableMap.Builder<String, String> mapBuilder = ImmutableMap.<String, String>builder()
```
ok, fixed.

zhaoamong commented on src/main/java/org/jenkinsci/plugins/argusnotifier/JenkinsRunFormatter.java:
```diff
@@ -55,6 +59,34 @@ String getRunUrl() {
+            }
+        }
+        catch (IOException | InterruptedException e) {
+            return "";
```
added a log info about an exception

zhaoamong commented:
Hi @justinharringa 

I hope the new tags I added just as property of a Jenkins job run duration metrics(build.time, total.build.time, queue.time). 
I added 3 tags including (git_commit, build_number, build_status), so it doesn't make sense to add a tag "build_status" to a metric named "build_status" in which "status" is both tag and datapoint.

I'd like to keep the change minimum. Do not have a big surface area that I can't test fully.

justinharringa commented:
I'm going to go ahead and merge and take care of the rest of my requested changes.
